domain/externalcontroller/state/state.go

```diff
@@ -127,57 +128,131 @@ func (st *State) UpdateExternalController(
 		return errors.Trace(err)
 	}
 
-	cID := ci.ControllerTag.Id()
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return st.updateExternalControllerTx(ctx, ci, modelUUIDs, tx)
+	})
+
+	return errors.Trace(err)
+}
+
+// ImportExternalControllers imports the list of MigrationControllerInfo
+// external controllers on one single transaction.
+func (st *State) ImportExternalControllers(ctx context.Context, infos []externalcontroller.MigrationControllerInfo) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		q := `
+		for _, ci := range infos {
+			err := st.updateExternalControllerTx(
+				ctx,
+				crossmodel.ControllerInfo{
+					ControllerTag: ci.ControllerTag,
+					Addrs:         ci.Addrs,
+					Alias:         ci.Alias,
+					CACert:        ci.CACert,
+				},
+				ci.ModelUUIDs,
+				tx,
+			)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func (st *State) updateExternalControllerTx(
+	ctx context.Context,
+	ci crossmodel.ControllerInfo,
+	modelUUIDs []string,
+	tx *sql.Tx,
```

Comment:
This argument should come before the data, and be adjacent to `ctx`.

Reply:
done